File: app/models/track.py
"""Track models — maps to DataFrame rows and API responses."""

from __future__ import annotations

from pydantic import BaseModel, field_validator


class TrackRow(BaseModel):
    """A single track as returned by the API.

    Maps to one row of the in-memory DataFrame, plus computed fields.
    Pydantic v2 handles numpy int64/float64 → Python int/float automatically.
    """

    id: int
    title: str
    artist: str
    album_title: str | None = None
    bpm: float | None = None
    key: str | None = None
    year: int | None = None
    comment: str | None = None
    location: str | None = None
    status: str | None = None  # "tagged" | "untagged"

    # Parsed facets (only included in detailed responses)
    genre1: str | None = None
    genre2: str | None = None
    descriptors: str | None = None
    mood: str | None = None
    track_location: str | None = None  # parsed location (avoids clash with file path)
    era: str | None = None

    @field_validator("bpm", mode="before")
    @classmethod
    def coerce_bpm(cls, v):
        """Handle NaN from pandas."""
        if v is None:
            return None
        try:
            f = float(v)
            if f != f:  # NaN check
                return None
            return f
        except (TypeError, ValueError):
            return None

    @field_validator("year", mode="before")
    @classmethod
    def coerce_year(cls, v):
        """Handle NaN and float years from pandas."""
        if v is None:
            return None
        try:
            f = float(v)
            if f != f:  # NaN check
                return None
            return int(f)
        except (TypeError, ValueError):
            return None


class TrackSearchResult(TrackRow):
    """Track with search scoring fields."""

    score: float | None = None
    matched: dict | None = None


class TrackExample(BaseModel):
    """Exemplar track shown in tree nodes — minimal fields."""

    title: str
    artist: str
    year: int | None = None

File: frontend/src/schemas/track.ts
import { z } from 'zod'

export const TrackRow = z.object({
  id: z.number(),
  title: z.string(),
  artist: z.string(),
  album_title: z.string().nullable().optional(),
  bpm: z.number().nullable().optional(),
  key: z.string().nullable().optional(),
  year: z.number().nullable().optional(),
  comment: z.string().nullable().optional(),
  location: z.string().nullable().optional(),
  status: z.string().nullable().optional(),
  genre1: z.string().nullable().optional(),
  genre2: z.string().nullable().optional(),
  descriptors: z.string().nullable().optional(),
  mood: z.string().nullable().optional(),
  track_location: z.string().nullable().optional(),
  era: z.string().nullable().optional(),
})

export const TrackSearchResult = TrackRow.extend({
  score: z.number().nullable().optional(),
  matched: z.record(z.string(), z.unknown()).nullable().optional(),
})

export const TagSingleResponse = z.object({
  id: z.number(),
  comment: z.string(),
  year: z.number().nullable().optional(),
})

export type TrackRow = z.infer<typeof TrackRow>
export type TrackSearchResult = z.infer<typeof TrackSearchResult>
export type TagSingleResponse = z.infer<typeof TagSingleResponse>
